File: leetcode3/최원준/174. Dungeon Game.py
#

'''
1. 아이디어 :
각 grid에 2가지 정보를 저장한다. 지금까지 온 경로의:
- 현재 체력이 제일 높은 경로의 [현재 체력, 최악의 체력]
- 지금까지 가장 덜 위험했던 경로의 [현재 체력, 최악의 체력]

매 grid를 방문할때마다 왼쪽의 ([현재 체력, 최악의 체력], [현재 체력, 최악의 체력])과 위의 ([현재 체력, 최악의 체력], [현재 체력, 최악의 체력])을 대상으로 계산한다.

2. 시간복잡도 :
    O(n*m*2)

3. 자료구조/알고리즘 :
누적합

'''

class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        n = len(dungeon)
        m = len(dungeon[0])

        grids = [[None for _ in range(m)] for _ in range(n)]

        initial = dungeon[0][0]
        grids[0][0] = ([initial, initial], [initial, initial])

        def update(path, value):
            cur, worst = path
            new_cur = cur + value
            new_worst = min(worst, new_cur)
            return [new_cur, new_worst]

        def select(paths):
            best_current = max(paths, key=lambda x: (x[0], x[1]))
            best_worst = max(paths, key=lambda x: (x[1], x[0]))
            return best_current, best_worst

        for i in range(n):
            for j in range(m):
                if i == 0 and j == 0:
                    continue

                candidates = []
                value = dungeon[i][j]

                if i > 0:
                    up_current, up_worst = grids[i - 1][j]
                    candidates.append(update(up_current, value))
                    candidates.append(update(up_worst, value))

                if j > 0:
                    left_current, left_worst = grids[i][j - 1]
                    candidates.append(update(left_current, value))
                    candidates.append(update(left_worst, value))

                grids[i][j] = select(candidates)

        end_current, end_worst = grids[n - 1][m - 1]

        return min(
            max(1, 1 - end_current[1]),
            max(1, 1 - end_worst[1])
        )

File: leetcode3/이진희/174_Dungeon Game.java
/*

1. 아이디어 : dp 활용
              오른쪽, 아래 방향으로만 이동 가능
              이동 도중 0 이하가 되면 안됨
              도착지부터 출발지까지 역방향으로 이동하며, 최소 Stat를 찾는다

              포인트: 원래 던전 크기보다 +1씩 배열값을 초기화
                      도착지 [m-1][n-1]근처는 1로 초기화 하여 계산 (경계값 처리 줄이기)

2. 시간복잡도 : O(M * N)

3. 자료구조/알고리즘 : 이차원  DP

 */

class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        int m = dungeon.length;
        int n = dungeon[0].length;

        int[][] dp = new int[m+1][n+1];
        for(int i = 0; i <= m; i++) {
            for(int j = 0; j <= n; j++) {
                dp[i][j] = Integer.MAX_VALUE;
            }
        }

        dp[m][n-1] = 1;
        dp[m-1][n] = 1;

        for(int i=m-1; i>=0; i--) {
            for(int j=n-1; j>=0; j--) {
                int a = dp[i][j+1];
                int b = dp[i+1][j];
                int status = dungeon[i][j];

                dp[i][j] = Math.max(1, Math.min(a, b) - status);
            }
        }

        return dp[0][0];
    }
}